src/rdkafka_request.c

```diff
@@ -4773,12 +4900,313 @@ int rd_kafka_ProduceRequest(rd_kafka_broker_t *rkb,
+                                                                of_part_cnt, part_cnt);
+                                }
+                                /* Write the topic name */
+                                rd_kafka_buf_write_kstr(request_rkbuf, cur_topic->rkt_topic);
+                                /* Partition Count: updated later */
+                                of_part_cnt = rd_kafka_buf_write_arraycnt_pos(request_rkbuf);
+                                part_cnt = 0;
+                                topic_cnt++;
+                        }
+
+                        rd_kafka_toppar_t *rktp = cur_rkbuf->rkbuf_batch.rktp;
+
+                        /* Write the partition */
+                        rd_kafka_buf_write_i32(request_rkbuf, rktp->rktp_partition);
+
+                        /* Copy the already encoded/compressed RecordBatch from buf */
+                        rd_kafka_copy_batch_buf(cur_rkbuf, request_rkbuf);
+
+                        /* Move the msgbatch from the cur_rkbuf into the new request's
+                           batch_list. A new rd_kafka_msgbatch_t is created and we move
+                           the each msgs in the msgq over. This is so that we can reused
+                           as much as possible the existing memory accounting/cleanup.
+                           The actual message is not copied.
+                           This list of msgbatch will be cleaned up when the buf of this
+                           MultiBatch request is destroyed */
+                        rd_kafka_msgbatch_t *src = &cur_rkbuf->rkbuf_batch;
+                        rd_kafka_msgbatch_t *dst = rd_malloc(sizeof(rd_kafka_msgbatch_t));
+                        rd_kafka_msgbatch_init(dst, src->rktp, src->pid, src->epoch_base_msgid);
+                        rd_kafka_msgq_move(&dst->msgq, &src->msgq);
+                        rd_list_add(&request_rkbuf->rkbuf_u.Produce.batch_list, dst);
+
+                        part_cnt++;
+                        prev_topic = cur_topic;
+                        prev_rkbuf = cur_rkbuf;
+                }
+
+                /* Finalize last topic level encodings */
+                finalize_topic_encoding(request_rkbuf, prev_rkbuf, of_part_cnt, part_cnt);
+
+                /* Topic count */
+                rd_kafka_buf_finalize_arraycnt(request_rkbuf, of_topic_cnt, topic_cnt);
+
+                /* The timeout setting is copy/pasted from rd_kafka_ProduceRequest func
+                   Instead of taking the first msg from a batch, we took the first
+                   msg from each batch, and choose the closest one. */
+                now = rd_clock();
+                first_msg_timeout =
+                        get_first_msg_timeout(&request_rkbuf->rkbuf_u.Produce.batch_list);
+                if (unlikely(first_msg_timeout <= 0)) {
+                        /* Message has already timed out, allow 100 ms
+                        * to produce anyway */
```

Comment:
why?

Reply:
This is a copy/paste from existing logic in rd_kafka_ProduceRequest. I specified in the comments above.

Reply:
This is a copy/paste from existing logic in rd_kafka_ProduceRequest. I specified in the comments above.